### PR TITLE
MLIR#808: fixed rock-linalg-align for mixed types

### PR DESCRIPTION
**Pull request by sjw36**
* added I8 support for rock.threadwise_write_all
* updated fusion/quantization.mlir test to run kernel pipeline

https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/808

### CODE REVIEW

File: mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp [hunk 1]
```diff
@@ -96,24 +96,29 @@ static Value transformAccordingly(PatternRewriter &b, Value nOut, Value refOp) {
...
   }
 
   return nOut;
 }
 
+Value makeRegs(PatternRewriter &b, MemRefType::Builder &mrb, Location loc,
+               Type srcType) {
+  auto srcMemType = srcType.cast<MemRefType>();
+  // 1. create a second allocation of the same type to hold loaded elements
+  return b.create<GpuAllocOp>(loc, static_cast<MemRefType>(mrb.setElementType(
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:110]: Address space needs to be set here.

File: mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp [hunk 2]
```diff
@@ -183,44 +188,40 @@ static ThreadwiseWriteAllOp traceToThreadwiseWrite(Value inp) {
...
       newInput = applyTransforms(b, twWriteOp, inp);
     }
     newInputs.push_back(newInput);
-    laGenericAMaps.push_back(AffineMap::getMultiDimIdentityMap(
-        newInput.getType().template cast<MemRefType>().getRank(), ctx));
-  }
 
-  laGenericAMaps.push_back(
-      AffineMap::getMultiDimIdentityMap(regType.getRank(), ctx));
+    auto inpRank = newInput.getType().template cast<ShapedType>().getRank();
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp]: No need to say `template` here

File: mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp [hunk 3]
```diff
@@ -271,10 +272,10 @@ LAGenericRewritePattern::matchAndRewrite(linalg::GenericOp laGeneric,
   if (!laGenericInputLeadingToGemmStore)
     return failure();
 
-  auto actualLAGenericOut = laGeneric.getOutputOperand(0);
-  if (laGenericInputLeadingToGemmStore.getType() !=
-      actualLAGenericOut->get().getType()) {
-    // TODO(sjw): fix for mixed types
+  auto outType = out.getType().template cast<ShapedType>();
+  auto inpType =
+      laGenericInputLeadingToGemmStore.getType().template cast<ShapedType>();
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp]: Same nit note on the template bit

**Review by krzysz00 — CHANGES_REQUESTED** [review]
You're creating `GpuAlloc` ops without specifying the private address space. This can cause issues.

Also, you need to also update GlobalStore, BufferLoad, ... to list `I8`

File: mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp (hunk 1 above)
Comment (sjw36) [reply mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:110]: It gets that from the MemRefType::Builder which is initialized by the register type for the blockwise_gemm output.

File: mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp (hunk 2 above)
Comment (sjw36) [reply mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp]: yes, copied code.. it is ugly.

**Comment by sjw36** [comment]
> Also, you need to also update GlobalStore, BufferLoad, ... to list `I8`

Updated rock.global_store. buffer_load already had I8.

**Review by jerryyin — APPROVED** [review]
Nice work enabling the alignment of memref of different type but same shape. LGTM

**Review by krzysz00 — APPROVED** [review]
Looks good to me

**Pull request merged**